### PR TITLE
VM callbacks can now raise (and be caught, or show their stacktrace)

### LINKED ISSUE #12676: Nimscript listFiles should throw exception when path is not found
[`listFiles`](https://nim-lang.org/docs/nimscript.html#listFiles%2Cstring) will silently return no items if the path specified does not exist, this can make it seem like a command has worked when it in fact has not.

### Example
```nim
task render, "Renders SVGs to PNGs":
  for file in listFiles(getCurrentDir() / "bad_path"):
    discard
```

### Current Output
Nothing

### Expected Output
Exception

### PR DESCRIPTION
**Pull request by timotheecour**
## before PR
VM callbacks can't be caught, so code behaves differently at RT vs CT if VM code calls a callback that raises. eg:
* `catch` doesn't work
* on crash, no useful context or stacktrace is shown so we have no idea how we got here

## after PR
* VM callbacks can be caught, all that's needed is an extra arg `registerCallback(.., mayRaise=true)`
* the exception (whether caught in regular code or if there's no handler, when the program crashes with stacktrace), contains the relevant context
* stacktrace inside VM callback is also shown when nim itself is compiled with --stacktrace:on
* nimscript implementation of callbacks is now simplified thanks to this and shows more informative stacktraces, eg when nim is built with --stacktrace:on
* so things like `walkDir(..,checkDir=true)`, `writeFile`, `removeDir` etc now work correctly in VM
* fixes https://github.com/nim-lang/Nim/issues/12676 (`listFiles(..,checkDir=true)` + friends now work; I did not change the default semantic to be consistent with `walkDir`)
* a new way to define internal procs is defined (`{.internalproc.}`), analog to `{.compilerproc.}`, but for semantic phase instead of codegen. This enables writing handlers in pure nim code instead of having to manually write AST; this not only simplifies writing compiler code, but it also allows change those handlers **without having to recompile nim**. A key aspect is that these are **private** (such symbols don't need to be exported).

## example
```nim
proc fun1() = writeFile("/tmp/nonexistant/bar.txt", "foo")
proc fun2()=fun1()
static: fun2()
# note: try/catch would also work inside VM after this PR (unlike before PR)
```
before PR: no context shown, hard to debug in complex programs
```
/Users/timothee/git_clone/nim/Nim_devel/lib/system/io.nim(706) writeFile
Error: unhandled exception: cannot open: /tmp/nonexistant/bar.txt [IOError]
```

after PR:
```
stack trace: (most recent call last)
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10375c.nim(93, 15) t10375c
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10375c.nim(92, 19) fun2
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10375c.nim(91, 26) fun1
/Users/timothee/git_clone/nim/Nim_prs/lib/system.nim(3024, 9) nimInternalNewException
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10375c.nim(1, 1) template/generic instantiation from here
/Users/timothee/git_clone/nim/Nim_prs/lib/system.nim(3024, 9) Error: unhandled exception: cannot open: /tmp/nonexistant/bar.txt
VM callback stacktrace (compile nim with --stacktrace if needed)
/Users/timothee/git_clone/nim/Nim_prs/lib/system/io.nim(706) writeFile

 [IOError]
          raise newException(T, msg)
          ^
```

## stacktrace inside VM callback
furthermore, when nim itself is compiled with `--stacktace`, the msg will also show the VM stacktrace (truncated to the useful part) eg:
```
stack trace: (most recent call last)
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10375b.nims(18, 7)
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10375b.nims(17, 20) fun2
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10375b.nims(16, 7) fun
/Users/timothee/git_clone/nim/Nim_prs/lib/system/nimscript.nim(325, 16) cd
/Users/timothee/git_clone/nim/Nim_prs/lib/system.nim(3023, 9) nimInternalNewException
/Users/timothee/git_clone/nim/Nim_prs/lib/system.nim(3023, 9) Error: unhandled exception: No such file or directory
Additional info: "nonexistant"
VM callback stacktrace
/Users/timothee/git_clone/nim/Nim_prs/compiler/vm.nim(1231) rawExecute
/Users/timothee/git_clone/nim/Nim_prs/compiler/scriptconfig.nim(88) :anonymous
/Users/timothee/git_clone/nim/Nim_prs/lib/pure/os.nim(1367) setCurrentDir
/Users/timothee/git_clone/nim/Nim_prs/lib/pure/includes/oserr.nim(94) raiseOSError

 [OSError]
          raise newException(T, msg)
          ^
```

## TODO for future PR's
* after this PR, it should be easy to catch C++ exceptions from CT FFI
* handle similarly `opcNDynBindSym`

## links
* [using readFile at compile time break exceptions - Nim forum](https://forum.nim-lang.org/t/8367)

### CODE REVIEW

File: compiler/vm.nim [hunk 1]
```diff
@@ -540,6 +540,27 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
         "ra", regDescr("ra", ra), "rb", regDescr("rb", instr.regB),
         "rc", regDescr("rc", instr.regC)]
 
+    proc onProcCall(prc: PSym, isClosure: bool, rb = 0, rc = 0, regs2: var seq[TFullReg]) =
```
Comment (timotheecour) [inline compiler/vm.nim:537]: note to reviewer: this code blocks is mostly moved without change; the only change involves instances of `regs2` vs `regs` (1 `regs` left intentionally)

**Comment by krux02** [comment]
I can't find the test case.

**Comment by timotheecour** [comment]
done PTAL

File: compiler/pragmas.nim [hunk 2]
```diff
@@ -897,6 +897,12 @@ proc singlePragma(c: PContext, sym: PSym, n: PNode, i: var int,
         incl(sym.flags, sfWasForwarded)
       of wDynlib:
         processDynLib(c, it, sym)
+      of wInternalProc:
+        noVal(c, it)
+        if sym.name == getIdent(c.cache, "nimInternalNewException"):
```
Comment (krux02) [inline compiler/pragmas.nim]: you can do cases sensitive string comparison here as well.

File: lib/system.nim [hunk 3]
```diff
@@ -3007,3 +3006,25 @@ export io
...
+        var msg = msg
+        if traceMsg.len > 0:
+          # we could query for whether nim was build with --stacktrace to
+          # provide more accurate msg, but not sure if it's exposed.
+          msg.add "\nVM callback stacktrace (compile nim with --stacktrace if needed)\n" & traceMsg & "\n"
+        raise newException(T, msg)
+    # can add more hardcoded standard types here
+    fun(OSError)
+    fun(IOError)
+    fun(AssertionError)
```
Comment (krux02) [inline lib/system.nim:3032]: Does thins mean that only `OSError`, `IOError` and `AssertionError` can be raised? Isn't there space to have a generic solution here that works on all error kinds?

Comment (Clyybber) [inline lib/system.nim:3032]: This is horrible.

**Review by krux02 — CHANGES_REQUESTED** [review]
You are adding a new compiler feature: `internalproc`. Please make a separate PR for that feature alone. You can reference this PR there. I think that feature should be reviewed and tested independently.

This is not a full review yet. I just skimmed the diff.

File: compiler/pragmas.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/pragmas.nim]: done, thanks for pointing this out; all i had to do was use `id` field

**Comment by cooldome** [comment]
I have reviewed this PR and really like it. 
Only two comment I have:
-    I don't think you need to introduce `wInternalProc` pragma, you can use `compilerProc` pragma and 
    `proc getCompilerProc*(g: ModuleGraph; name: string): PSym` from `magicsys` instead.
-  You have introduced argument `mayRaise` that use needs to be populated by the user. I would consider checkings `{.raises[].}` effect of symbol proc type instead and automatically infer if wrapped proc can raise exception or not.

**Comment by cooldome** [comment]
I have checked again and i think it safe to assume that wrapped proc can always raise exception, little overhead is added for potential exception handling.

File: lib/system.nim (hunk 3 above)
Comment (cooldome) [reply lib/system.nim:3032]: Yes, this bit needs improvement. Can't you just pass already created object and raise it?

Comment (cooldome) [reply lib/system.nim:3032]: If it is not possible to use compilerProc here then it needs to be moved into compiler

File: compiler/pragmas.nim [hunk 4]
```diff
@@ -24,7 +24,7 @@ const
     ## common pragmas for declarations, to a good approximation
   procPragmas* = declPragmas + {FirstCallConv..LastCallConv,
     wMagic, wNoSideEffect, wSideEffect, wNoreturn, wNosinks, wDynlib, wHeader,
-    wCompilerProc, wNonReloadable, wCore, wProcVar, wVarargs, wCompileTime, wMerge,
+    wCompilerProc, wNonReloadable, wCore, wInternalProc, wProcVar, wVarargs, wCompileTime, wMerge,
```
Comment (cooldome) [inline compiler/pragmas.nim:27]: Important to avoid introducing new pragmas

**Review by cooldome — CHANGES_REQUESTED** [review]

**Comment by Araq** [comment]
The addition to system.nim is terrible and the feature itself is of questionable value. Yes, I know you will not stop until everything from os.nim is available in the VM, but I don't like it and the existing mechanism might suffice. But as others remarked the worst part is the new pragma and the system.nim additions neither of which are required afaict.

**Pull request closed without merging**